File: scripts/status.py
#!/usr/bin/env python3
"""
status.py — /harness:status (F-005) 구현. Read-only, CQS.

사용:
  python3 scripts/status.py                           # cwd .harness 기준 human 출력
  python3 scripts/status.py --harness-dir DIR
  python3 scripts/status.py --json                    # 기계 파싱용
  python3 scripts/status.py --feature F-003           # 단일 피처 상세

CQS 불변조건: 파일 **수정 없음**. mtime 변경 안 함 (테스트 검증).

출력 섹션:
  1. session — active_feature · last_command · last_gate_passed
  2. feature counts (planned/in_progress/blocked/done/archived)
  3. drift_status (harness.yaml 에서)
  4. recent evidence (feature 별 최근 1~3 개)
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    print("pyyaml is required", file=sys.stderr)
    sys.exit(1)

_THIS = Path(__file__).resolve().parent
if str(_THIS) not in sys.path:
    sys.path.insert(0, str(_THIS))

from state import State  # noqa: E402


@dataclass
class StatusReport:
    session: dict
    counts: dict
    drift_status: str
    last_sync: dict | None
    features_summary: list[dict]
    active_feature: dict | None


def _load_harness_yaml(harness_dir: Path) -> dict:
    path = harness_dir / "harness.yaml"
    if not path.is_file():
        return {}
    with path.open("r", encoding="utf-8") as f:
        data = yaml.safe_load(f)
    return data if isinstance(data, dict) else {}


def _tail_events(harness_dir: Path, n: int = 1) -> list[dict]:
    """events.log 의 마지막 n 개를 JSON 디코드해서 반환. 실패 라인은 skip."""
    path = harness_dir / "events.log"
    if not path.is_file():
        return []
    lines = path.read_text(encoding="utf-8").splitlines()[-n:]
    out = []
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return out


def build_report(harness_dir: Path, *, feature_filter: str | None = None) -> StatusReport:
    """state.yaml + harness.yaml + events.log 를 읽어 StatusReport 생성."""
    st = State.load(harness_dir)
    harness_yaml = _load_harness_yaml(harness_dir)

    drift_status = harness_yaml.get("generation", {}).get("drift_status", "unknown")

    last_sync_events = _tail_events(harness_dir, n=5)
    last_sync = None
    for ev in reversed(last_sync_events):
        if ev.get("type") == "sync_completed":
            last_sync = {
                "ts": ev.get("ts"),
                "spec_hash": ev.get("spec_hash", "")[:12],
                "plugin_version": ev.get("plugin_version"),
            }
            break

    counts = st.feature_counts()

    # feature 요약 — filter 가 있으면 그 하나만, 없으면 전체 간단히
    features_summary: list[dict] = []
    for f in st.data["features"]:
        if not isinstance(f, dict):
            continue
        fid = f.get("id", "?")
        if feature_filter and fid != feature_filter:
            continue
        gates = f.get("gates", {}) or {}
        passed = [g for g, v in gates.items() if isinstance(v, dict) and v.get("last_result") == "pass"]
        failed = [g for g, v in gates.items() if isinstance(v, dict) and v.get("last_result") == "fail"]
        features_summary.append(
            {
                "id": fid,
                "status": f.get("status", "planned"),
                "started_at": f.get("started_at"),
                "completed_at": f.get("completed_at"),
                "gates_passed": passed,
                "gates_failed": failed,
                "evidence_count": len(f.get("evidence", []) or []),
            }
        )

    active_fid = st.data["session"].get("active_feature_id")
    active_feature = next(
        (f for f in features_summary if f["id"] == active_fid), None
    ) if active_fid else None

    return StatusReport(
        session=dict(st.data["session"]),
        counts=counts,
        drift_status=drift_status,
        last_sync=last_sync,
        features_summary=features_summary,
        active_feature=active_feature,
    )


def format_human(report: StatusReport) -> str:
    lines = ["📋 /harness:status"]
    lines.append("")

    s = report.session
    lines.append(f"Session")
    lines.append(f"  started_at         {s.get('started_at') or '—'}")
    lines.append(f"  last_command       {s.get('last_command') or '—'}")
    lines.append(f"  last_gate_passed   {s.get('last_gate_passed') or '—'}")
    lines.append(f"  active_feature_id  {s.get('active_feature_id') or '—'}")
    lines.append("")

    c = report.counts
    total = sum(c.values())
    lines.append(f"Features ({total})")
    for st, n in c.items():
        if n:
            lines.append(f"  {st:12} {n}")
    lines.append("")

    lines.append(f"Drift status: {report.drift_status}")
    lines.append("")

    if report.last_sync:
        ls = report.last_sync
        lines.append(f"Last sync: {ls['ts']} · spec_hash={ls['spec_hash']} · plugin={ls['plugin_version']}")
        lines.append("")

    if report.active_feature:
        af = report.active_feature
        lines.append(f"Active feature: {af['id']} [{af['status']}]")
        if af.get("gates_passed"):
            lines.append(f"  gates passed: {', '.join(af['gates_passed'])}")
        if af.get("gates_failed"):
            lines.append(f"  gates failed: {', '.join(af['gates_failed'])}")
        lines.append(f"  evidence: {af['evidence_count']} entries")
        lines.append("")

    if not report.active_feature and report.features_summary:
        # filter 가 없을 때는 간단한 표만, 없으면 skip
        pass

    return "\n".join(lines).rstrip() + "\n"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="/harness:status (F-005) — read-only state view")
    parser.add_argument(
        "--harness-dir",
        type=Path,
        default=Path.cwd() / ".harness",
    )
    parser.add_argument("--feature", default=None, help="restrict to a single feature id")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args(argv)

    if not args.harness_dir.is_dir():
        print(f"error: {args.harness_dir} not found", file=sys.stderr)
        return 2

    report = build_report(args.harness_dir, feature_filter=args.feature)

    if args.json:
        json.dump(
            {
                "session": report.session,
                "counts": report.counts,
                "drift_status": report.drift_status,
                "last_sync": report.last_sync,
                "features": report.features_summary,
                "active_feature": report.active_feature,
            },
            sys.stdout,
            indent=2,
            ensure_ascii=False,
        )
        print()
    else:
        sys.stdout.write(format_human(report))

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/status.ts
/**
 * `/harness:status` read-only summary (F-094 port of
 * `scripts/status.py`, F-005 in spec).
 *
 * CQS contract — never modifies any file under the harness
 * directory; mtime invariant is enforced by tests.
 *
 * Composes three reads:
 *
 *   - {@link import('./core/state.ts').State} — feature counts,
 *     session block, evidence summary.
 *   - `harness.yaml` — drift_status from `generation.drift_status`.
 *   - `events.log` tail — last `sync_completed` event for the
 *     last_sync block.
 *
 * @module status
 */

import {readFileSync, statSync} from 'node:fs';
import {join} from 'node:path';
import {parse as yamlParse} from 'yaml';

import {State, type StateData} from './core/state.js';

/** Per-feature short summary (cards in the human view). */
export interface FeatureSummary {
  id: string;
  status: string;
  started_at: string | null;
  completed_at: string | null;
  gates_passed: string[];
  gates_failed: string[];
  evidence_count: number;
}

/** Optional last_sync block when an event has been observed. */
export interface LastSyncSummary {
  ts: string | null;
  spec_hash: string;
  plugin_version: string | null;
}

/** Full status report shape; both `formatHuman` and JSON output share it. */
export interface StatusReport {
  session: StateData['session'];
  counts: Record<string, number>;
  drift_status: string;
  last_sync: LastSyncSummary | null;
  features_summary: FeatureSummary[];
  active_feature: FeatureSummary | null;
}

/** Optional input for {@link buildReport}. */
export interface BuildReportOptions {
  featureFilter?: string | null;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function loadHarnessYaml(harnessDir: string): Record<string, unknown> {
  const path = join(harnessDir, 'harness.yaml');
  try {
    if (!statSync(path).isFile()) {
      return {};
    }
  } catch {
    return {};
  }
  const raw = readFileSync(path, 'utf-8');
  const parsed: unknown = yamlParse(raw);
  return isPlainObject(parsed) ? parsed : {};
}

/** Reads the last n JSON-line entries from `events.log`. */
function tailEvents(harnessDir: string, n: number = 1): Array<Record<string, unknown>> {
  const path = join(harnessDir, 'events.log');
  try {
    if (!statSync(path).isFile()) {
      return [];
    }
  } catch {
    return [];
  }
  const raw = readFileSync(path, 'utf-8');
  const lines = raw.split('\n').filter((line) => line.trim().length > 0);
  const tail = lines.slice(-n);
  const out: Array<Record<string, unknown>> = [];
  for (const line of tail) {
    try {
      const parsed: unknown = JSON.parse(line);
      if (isPlainObject(parsed)) {
        out.push(parsed);
      }
    } catch {
      continue;
    }
  }
  return out;
}

/** Builds a {@link StatusReport} for the given harness directory. */
export function buildReport(
  harnessDir: string,
  options: BuildReportOptions = {},
): StatusReport {
  const featureFilter = options.featureFilter ?? null;
  const state = State.load(harnessDir);
  const harnessYaml = loadHarnessYaml(harnessDir);

  const generation = isPlainObject(harnessYaml['generation'])
    ? (harnessYaml['generation'] as Record<string, unknown>)
    : {};
  const driftStatus = typeof generation['drift_status'] === 'string'
    ? generation['drift_status']
    : 'unknown';

  const recentEvents = tailEvents(harnessDir, 5);
  let lastSync: LastSyncSummary | null = null;
  for (let i = recentEvents.length - 1; i >= 0; i--) {
    const ev = recentEvents[i]!;
    if (ev['type'] === 'sync_completed') {
      const ts = typeof ev['ts'] === 'string' ? ev['ts'] : null;
      const specHash = typeof ev['spec_hash'] === 'string' ? ev['spec_hash'].slice(0, 12) : '';
      const pluginVersion =
        typeof ev['plugin_version'] === 'string' ? ev['plugin_version'] : null;
      lastSync = {ts, spec_hash: specHash, plugin_version: pluginVersion};
      break;
    }
  }

  const counts: Record<string, number> = state.featureCounts();

  const featuresSummary: FeatureSummary[] = [];
  for (const f of state.data.features) {
    if (!isPlainObject(f)) {
      continue;
    }
    const fid = typeof f['id'] === 'string' ? f['id'] : '?';
    if (featureFilter !== null && fid !== featureFilter) {
      continue;
    }
    const gates = isPlainObject(f['gates']) ? (f['gates'] as Record<string, unknown>) : {};
    const passed: string[] = [];
    const failed: string[] = [];
    for (const [g, v] of Object.entries(gates)) {
      if (!isPlainObject(v)) {
        continue;
      }
      const result = v['last_result'];
      if (result === 'pass') {
        passed.push(g);
      } else if (result === 'fail') {
        failed.push(g);
      }
    }
    featuresSummary.push({
      id: fid,
      status: typeof f['status'] === 'string' ? f['status'] : 'planned',
      started_at: typeof f['started_at'] === 'string' ? f['started_at'] : null,
      completed_at: typeof f['completed_at'] === 'string' ? f['completed_at'] : null,
      gates_passed: passed,
      gates_failed: failed,
      evidence_count: Array.isArray(f['evidence']) ? (f['evidence'] as unknown[]).length : 0,
    });
  }

  const activeFid = state.data.session.active_feature_id;
  const activeFeature =
    typeof activeFid === 'string'
      ? featuresSummary.find((f) => f.id === activeFid) ?? null
      : null;

  return {
    session: {...state.data.session},
    counts,
    drift_status: driftStatus,
    last_sync: lastSync,
    features_summary: featuresSummary,
    active_feature: activeFeature,
  };
}

/** Trims trailing whitespace, mirroring Python's `str.rstrip()`. */
function rstrip(s: string): string {
  return s.replace(/\s+$/, '');
}

function pad(s: string, width: number): string {
  return s.length >= width ? s : s + ' '.repeat(width - s.length);
}

/**
 * Renders a {@link StatusReport} as the human view (matching Python
 * `format_human`).
 */
export function formatHuman(report: StatusReport): string {
  const lines: string[] = ['📋 /harness:status', ''];

  const s = report.session;
  lines.push('Session');
  lines.push(`  started_at         ${s.started_at ?? '—'}`);
  lines.push(`  last_command       ${s.last_command || '—'}`);
  lines.push(`  last_gate_passed   ${s.last_gate_passed ?? '—'}`);
  lines.push(`  active_feature_id  ${s.active_feature_id ?? '—'}`);
  lines.push('');

  const c = report.counts;
  const total = Object.values(c).reduce((acc, n) => acc + n, 0);
  lines.push(`Features (${total})`);
  for (const [st, n] of Object.entries(c)) {
    if (n > 0) {
      lines.push(`  ${pad(st, 12)} ${n}`);
    }
  }
  lines.push('');

  lines.push(`Drift status: ${report.drift_status}`);
  lines.push('');

  if (report.last_sync !== null) {
    const ls = report.last_sync;
    lines.push(
      `Last sync: ${ls.ts ?? ''} · spec_hash=${ls.spec_hash} · plugin=${ls.plugin_version ?? ''}`,
    );
    lines.push('');
  }

  if (report.active_feature !== null) {
    const af = report.active_feature;
    lines.push(`Active feature: ${af.id} [${af.status}]`);
    if (af.gates_passed.length > 0) {
      lines.push(`  gates passed: ${af.gates_passed.join(', ')}`);
    }
    if (af.gates_failed.length > 0) {
      lines.push(`  gates failed: ${af.gates_failed.join(', ')}`);
    }
    lines.push(`  evidence: ${af.evidence_count} entries`);
    lines.push('');
  }

  return `${rstrip(lines.join('\n'))}\n`;
}
